paddle/framework/executor.h

```diff
@@ -14,19 +14,100 @@ limitations under the License. */
 
 #pragma once
 
+#include <map>
+#include <unordered_map>
+
 #include "paddle/framework/op_info.h"
 #include "paddle/framework/program_desc.h"
 #include "paddle/framework/scope.h"
 #include "paddle/framework/tensor.h"
+#include "paddle/platform/device_context.h"
 
 namespace paddle {
 namespace framework {
 
+class DeviceContextPool {
+ public:
+  static DeviceContextPool& Get() {
+    PADDLE_ENFORCE_NOT_NULL(pool, "Need to Create DeviceContextPool first!");
```

Comment:
Can create if `nullptr` here. So the `Create` function is not neded.

Reply:
Yes, a singleton ONLY always mix these two function together, but here we have a slightly problem. 

Create will create the **all the devices**. But when we need get always only need create part of it. 
For example, Init function create all the available GPU, CPU devices, when we want to create two or more executors, then Get() happens part may not have all the device information.

In previous implementation, we definitly have one and only one executor, but now it is different since executor does not contain resources, we may have two or more executors.

Reply:
That makes sense.